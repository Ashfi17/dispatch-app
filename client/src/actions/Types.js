export const CREATE_USER = "CREATE_USER";
export const LOGIN = "LOGIN";
export const ERROR_MESSAGE = "ERROR_MESSAGE";

export const GET_DISPATCH_INFO = "GET_DISPATCH_INFO";
export const SEARCH = "SEARCH";
export const EXPORT = "EXPORT";
