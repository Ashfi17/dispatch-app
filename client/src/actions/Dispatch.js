import {
  GET_DISPATCH_INFO,
  SEARCH,
  EXPORT,
  ADD_SOURCE_DESTINATION_TRANSPORTER,
  GET_DESTINATION_CODE,
  GET_SOURCE_CODE,
  GET_TRANSPORTER_CODE,
  ADD_NEW_DISPATCH,
  ERROR_DISPATCH_MESSAGE,
} from "./Types";
import axios from "axios";
import config from "../config";
const url = config.url + "dispatch";

//function to get the details based on sort input
export const getDispatchInfo = (order) => (dispatch) => {
  return axios
    .post(url + "/sort-by-date", order, {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      // console.log()
      dispatch({
        type: GET_DISPATCH_INFO,
        payload: res.data.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};
//function to search by source code
export const search = (searchTerm) => (dispatch) => {
  return axios
    .post(url + "/search", searchTerm, {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      console.log(res.data.data);
      dispatch({
        type: SEARCH,
        payload: res.data.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};
//function to export excel sheet
export const exportExcel = () => (dispatch) => {
  return axios
    .get(url + "/get-excel", {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: EXPORT,
        payload: res.data.message,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};

//++++++++  for Importing new data ++++++++

// action for adding new source destination and transporter
export const addNewSDT = (data) => (dispatch) => {
  console.log("action called");
  return axios
    .post(url + "/add-source-destination", data, {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: ADD_SOURCE_DESTINATION_TRANSPORTER,
        payload: res.data.message,
      });
    })
    .catch((err) => {
      dispatch({
        type: ERROR_DISPATCH_MESSAGE,
        payload: err.response.data.message,
      });
    });
};

//action for getting all the source codes
export const getSourceCode = () => (dispatch) => {
  return axios
    .get(url + "/source_code", {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: GET_SOURCE_CODE,
        payload: res.data.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};
//action for getting all the destination codes
export const getDestinationCode = () => (dispatch) => {
  return axios
    .get(url + "/destination_code", {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: GET_DESTINATION_CODE,
        payload: res.data.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};

//action for getting all the transporter codes
export const getTransporterCode = () => (dispatch) => {
  return axios
    .get(url + "/transporter_code", {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: GET_TRANSPORTER_CODE,
        payload: res.data.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};
export const addNewDispatch = (dispatchData) => (dispatch) => {
  return axios
    .post(url + "/add-dispatch-details", dispatchData, {
      headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
    })
    .then((res) => {
      dispatch({
        type: ADD_NEW_DISPATCH,
        payload: res.data.message,
      });
    })
    .catch((err) => {
      dispatch({
        type: ERROR_DISPATCH_MESSAGE,
        payload: err.response.data.message,
      });
    });
};
