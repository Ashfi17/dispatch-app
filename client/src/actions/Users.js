import { CREATE_USER, LOGIN, ERROR_MESSAGE } from "./Types";
import config from "../config";
import axios from "axios";
const url = config.url + "users";

export const createUser = (userDetail, history) => (dispatch) => {
  return axios
    .post(url + "/new", userDetail)
    .then((res) => {
      dispatch({
        type: CREATE_USER,
      });
      history.push("/login");
    })
    .catch((err) => {
      // console.log(err.response.data.message);
      dispatch({ type: ERROR_MESSAGE, payload: err.response.data.message });
    });
};
export const login = (userCredentials, history) => (dispatch) => {
  return axios
    .post(url + "/login", userCredentials)
    .then((res) => {
      localStorage.setItem("token", res.data.token);
      dispatch({
        type: LOGIN,
      });
      history.push("/");
    })
    .catch((err) => {
      dispatch({ type: ERROR_MESSAGE, payload: err.response.data.message });
    });
};
