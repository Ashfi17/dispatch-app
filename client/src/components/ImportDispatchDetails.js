import React, { Component } from "react";
import jwt_decode from "jwt-decode";
import "../assets/styles/ImportDetails.css";
import {
  addNewSDT,
  getSourceCode,
  getDestinationCode,
  getTransporterCode,
  addNewDispatch,
} from "../actions/Dispatch";
import { connect } from "react-redux";

export class ImportDispatchDetails extends Component {
  state = {
    source: "",
    destination: "",
    transporter: "",
    vehicleNumber: "",
    startDate: "",
    endDate: "",
    sourceValue: "",
    destinationValue: "",
    transporterValue: "",
    sourceValueFlag: false,
    destinationValueFlag: false,
    transporterValueFlag: false,
    errorMessage: "",
    successMessage: "",
  };

  componentDidMount() {
    if (localStorage.getItem("token")) {
      var decode = jwt_decode(localStorage.getItem("token"));

      //checking if the token is expired or not
      if (Date.now() > decode.exp * 1000) {
        //if token is expired push to login page
        this.props.history.push("/login", {
          message: "Token expired please login again",
        });
      }
    } else {
      this.props.history.push("/login", {
        message: "No token present, please login",
      });
    }
  }

  OnTextChange = (event) => {
    this.setState({
      errorMessage: "",
      [event.target.name]: event.target.value,
    });
    if (event.target.name === "sourceValue") {
      if (event.target.value === "") this.setState({ sourceValueFlag: false });
    }
    if (event.target.name === "destinationValue") {
      if (event.target.value === "")
        this.setState({ destinationValueFlag: false });
    }
    if (event.target.name === "transporterValue") {
      if (event.target.value === "")
        this.setState({ transporterValueFlag: false });
    }
  };

  saveData = () => {
    var decode = jwt_decode(localStorage.getItem("token"));

    if (Date.now() > decode.exp * 1000) {
      this.props.history.push("/login", {
        message: "Token expired please login again",
      });
    } else {
      let data = {
        source: this.state.source,
        destination: this.state.destination,
        transporter: this.state.transporter,
      };
      this.props.addNewSDT(data);

      this.setState({
        source: "",
        destination: "",
        transporter: "",
        errorMessage: "Added Successfully",
      });
    }
  };

  saveDispatchDetails = async () => {
    const {
      sourceValue,
      destinationValue,
      transporterValue,
      vehicleNumber,
      startDate,
      endDate,
    } = this.state;
    var decode = jwt_decode(localStorage.getItem("token"));

    if (Date.now() > decode.exp * 1000) {
      this.props.history.push("/login", {
        message: "Token expired please login again",
      });
    } else {
      let dispatchData = {
        source_code: sourceValue,
        destination_code: destinationValue,
        transporter_code: transporterValue,
        start_date: startDate,
        end_date: endDate,
        vehicle_number: vehicleNumber,
      };

      this.props.addNewDispatch(dispatchData);
      this.setState({
        sourceValue: "",
        destinationValue: "",
        transporterValue: "",
        startDate: "",
        endDate: "",
        vehicleNumber: "",
      });
    }
  };

  addCode = (code, type) => {
    if (type === "source")
      this.setState({ sourceValue: code, sourceValueFlag: false });
    if (type === "destination")
      this.setState({ destinationValue: code, destinationValueFlag: false });
    if (type === "transporter")
      this.setState({ transporterValue: code, transporterValueFlag: false });
  };

  handleFocus = (event) => {
    var decode = jwt_decode(localStorage.getItem("token"));
    if (event.target.name === "sourceValue") {
      //checking is the token is expired
      if (Date.now() > decode.exp * 1000) {
        this.props.history.push("/login", {
          message: "Token expired please login again",
        });
      } else {
        this.props.getSourceCode();
        this.setState({ sourceValueFlag: true });
      }
    }
    if (event.target.name === "destinationValue") {
      if (Date.now() > decode.exp * 1000) {
        this.props.history.push("/login", {
          message: "Token expired please login again",
        });
      } else {
        this.props.getDestinationCode();
        this.setState({ destinationValueFlag: true });
      }
    }
    if (event.target.name === "transporterValue") {
      if (Date.now() > decode.exp * 1000) {
        this.props.history.push("/login", {
          message: "Token expired please login again",
        });
      } else {
        this.props.getTransporterCode();
        this.setState({ transporterValueFlag: true });
      }
    }
  };

  handleBlur = (target) => {
    if (target === "sourceValue") this.setState({ sourceValueFlag: false });
    if (target === "destinationValue")
      this.setState({ destinationValueFlag: false });
    if (target === "transporterValue")
      this.setState({ transporterValueFlag: false });
  };
  render() {
    return (
      <div>
        <div className="add-source-destination">
          <h1>Add new source,destination and transporter</h1>
          <input
            id="source-id"
            name="source"
            className="data-input-field"
            placeholder="source name"
            type="text"
            value={this.state.source}
            onChange={(e) => this.OnTextChange(e)}
          />
          <input
            id="destination-id"
            name="destination"
            className="data-input-field"
            placeholder="destination name"
            type="text"
            value={this.state.destination}
            onChange={(e) => this.OnTextChange(e)}
          />
          <input
            id="transporter-id"
            name="transporter"
            className="data-input-field"
            placeholder="transporter name"
            type="text"
            value={this.state.transporter}
            onChange={(e) => this.OnTextChange(e)}
          />
          <button
            id="add-details-submit-button"
            className="submit-button"
            onClick={this.saveData}
          >
            Save
          </button>
          <span className="error-disptach-text">
            {this.props.errordispatchMessage
              ? this.props.errordispatchMessage
              : this.props.DataMessage
              ? this.props.DataMessage
              : null}
          </span>
        </div>
        <h1 style={{ textAlign: "center" }}>Add New Dispatch Details</h1>
        <div className="add-new-dispatch-info">
          <div className="left-card-section">
            <input
              id="sourceValue-id"
              type="text"
              value={this.state.sourceValue}
              className="data-input-field"
              name="sourceValue"
              onChange={(e) => this.OnTextChange(e)}
              placeholder="Source code eg:S001"
              onFocus={(e) => this.handleFocus(e)}
            />
            {this.state.sourceValueFlag === true ? (
              <div id="code-list-id" className="codeList">
                {this.props.sourceCode.map((sources, key) => (
                  <span
                    id={"source-item" + key}
                    className="list-items"
                    key={key}
                    onClick={() => {
                      this.addCode(sources.source_code, "source");
                    }}
                    onBlur={() => this.handleBlur("sourceValue")}
                  >
                    {sources.source_code}
                  </span>
                ))}
              </div>
            ) : null}

            <input
              id="destinationValue-id"
              type="text"
              value={this.state.destinationValue}
              className="data-input-field right-section-input"
              name="destinationValue"
              onChange={this.OnTextChange}
              placeholder="Destination code eg:D001"
              onFocus={(e) => this.handleFocus(e)}
            />
            {this.state.destinationValueFlag === true ? (
              <div id="code-list-id" className="codeList">
                {this.props.destinationCode.map((destn, key) => (
                  <span
                    className="list-items"
                    id={"destination-item" + key}
                    key={key}
                    onClick={() => {
                      this.addCode(destn.destn_code, "destination");
                    }}
                    onBlur={() => this.handleBlur("destinationValue")}
                  >
                    {destn.destn_code}
                  </span>
                ))}
              </div>
            ) : null}
            <input
              type="text"
              id="transporterValue-id"
              value={this.state.transporterValue}
              className="data-input-field right-section-input"
              name="transporterValue"
              onChange={this.OnTextChange}
              placeholder="Transporter code eg:T001"
              onFocus={(e) => this.handleFocus(e)}
            />
            {this.state.transporterValueFlag === true ? (
              <div id="code-list-id" className="codeList">
                {this.props.transporterCode.map((transporter, key) => (
                  <span
                    key={key}
                    className="list-items"
                    id={"transporter-item" + key}
                    onClick={() => {
                      this.addCode(transporter.transporter_code, "transporter");
                    }}
                    onBlur={() => this.handleBlur("transporterValue")}
                  >
                    {transporter.transporter_code}
                  </span>
                ))}
              </div>
            ) : null}
          </div>
          <div className="right-card-section">
            <input
              id="vehicleNumber-id"
              name="vehicleNumber"
              placeholder="vehicle number"
              className="data-input-field "
              type="text"
              value={this.state.vehicleNumber}
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="startDate-id"
              name="startDate"
              placeholder="start Date"
              type="text"
              className="data-input-field right-section-input"
              value={this.state.startDate}
              onChange={(e) => this.OnTextChange(e)}
            />
            <input
              id="endDate-id"
              name="endDate"
              className="data-input-field right-section-input"
              placeholder="end Date"
              type="text"
              value={this.state.endDate}
              onChange={(e) => this.OnTextChange(e)}
            />
          </div>
        </div>

        <button
          id="add-dispatch-details-button"
          className="save-dispatch-button"
          onClick={this.saveDispatchDetails}
        >
          Add
        </button>
        <span
          className="error-disptach-text"
          style={{ paddingBottom: "30px", textAlign: "center" }}
        >
          {this.props.errordispatchMessage
            ? this.props.errordispatchMessage
            : this.props.DispatchInfoMessage
            ? this.props.DispatchInfoMessage
            : null}
        </span>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  excelMessage: state.DispatchReducer.excelMessage,
  sourceCode: state.DispatchReducer.sourceCode,
  destinationCode: state.DispatchReducer.destinationCode,
  transporterCode: state.DispatchReducer.transporterCode,
  DispatchInfoMessage: state.DispatchReducer.DispatchInfoMessage,
  errordispatchMessage: state.DispatchReducer.errordispatchMessage,
  DataMessage: state.DispatchReducer.DataMessage,
});
export default connect(mapStateToProps, {
  addNewSDT,
  getSourceCode,
  getDestinationCode,
  getTransporterCode,
  addNewDispatch,
})(ImportDispatchDetails);
