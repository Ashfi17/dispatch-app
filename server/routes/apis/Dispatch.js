const express = require("express");
const router = express.Router();
const mySqlConnection = require("../../connection");
const officegen = require("officegen");
const fs = require("fs");
const path = require("path");

//fetching all the source code
router.get("/source_code", (req, res, next) => {
  try {
    mySqlConnection.query(
      `SELECT source_code  FROM source_master  `,
      (err, result) => {
        if (!err) {
          res.status(200).send({
            status: 200,
            data: result,
            message: "fetched source_codes",
          });
        } else {
          res.status(400).send({
            status: 400,
            message: "Error fetching source  codes",
          });
        }
      }
    );
  } catch (e) {
    next(e);
  }
});
//fetching all the destination code
router.get("/destination_code", (req, res, next) => {
  try {
    mySqlConnection.query(
      `SELECT destn_code  FROM destination_master  `,
      (err, result) => {
        if (!err) {
          res.status(200).send({
            status: 200,
            data: result,
            message: "fetched destination_codes",
          });
        } else {
          res.status(400).send({
            status: 400,
            message: "Error fetching destination  codes",
          });
        }
      }
    );
  } catch (e) {
    next(e);
  }
});
//fetching all the transporter code
router.get("/transporter_code", (req, res, next) => {
  try {
    mySqlConnection.query(
      `SELECT  transporter_code from transporter_master   `,
      (err, result) => {
        if (!err) {
          res.status(200).send({
            status: 200,
            data: result,
            message: "fetched transporter_code",
          });
        } else {
          res.status(400).send({
            status: 400,
            message: "Error fetching transporter  codes",
          });
        }
      }
    );
  } catch (e) {
    next(e);
  }
});

//for adding source destination and transport details in master table
router.post("/add-source-destination", async (req, res, next) => {
  // toUpperCase()
  const source = req.body.source;
  const destination = req.body.destination;
  const transporter = req.body.transporter;
  try {
    if (!source || !destination) {
      res.status(400).send({
        message: "missing parameters",
      });
    } else {
      mySqlConnection.query(
        `(SELECT source_name  FROM source_master WHERE source_name = ?) UNION(SELECT destn_name FROM destination_master WHERE destn_name = ?)UNION(SELECT transporter_name FROM transporter_master WHERE transporter_name = ?)`,
        [
          source.toUpperCase(),
          destination.toUpperCase(),
          transporter.toUpperCase(),
        ],
        (err, result) => {
          if (!err) {
            if (result.length == 0) {
              mySqlConnection.query(
                "INSERT INTO source_master(source_name)VALUES(?)",
                [source.toUpperCase()],

                (err, result) => {
                  if (!err) {
                    mySqlConnection.query(
                      "INSERT INTO destination_master(destn_name)VALUES(?)",
                      [destination.toUpperCase()],
                      (err, result) => {
                        if (!err) {
                          mySqlConnection.query(
                            "INSERT INTO transporter_master(transporter_name)VALUES(?)",
                            [transporter.toUpperCase()],
                            (err, result) => {
                              if (!err) {
                                res.status(200).send({
                                  status: 200,
                                  message: "added data successfully",
                                });
                              }
                            }
                          );
                        } else {
                          res.status(400).send({
                            message: "Bad Request",
                          });
                        }
                      }
                    );
                  } else {
                    res.status(400).send({
                      message: "Bad Request",
                    });
                  }
                }
              );
            } else {
              res.status(200).send({
                message: "source /destination /transporte already exists",
              });
            }
          } else {
            res.status(400).send({
              message: "error",
            });
          }
        }
      );
    }
  } catch (e) {
    next(e);
  }
});

//for adding the dispatch details
router.post("/add-dispatch-details", (req, res, next) => {
  const sourceCode = req.body.source_code;
  const destinationCode = req.body.destination_code;
  const transporterCode = req.body.transporter_code;

  try {
    const startDate = new Date(
      req.body.start_date.split("-").reverse().join("-")
    )
      .toISOString()
      .slice(0, 10);
    const endDate = new Date(req.body.end_date.split("-").reverse().join("-"))
      .toISOString()
      .slice(0, 10);

    const vehicleNumber = req.body.vehicle_number;
    const vRegex = /(([A-Za-z]){2,3}(|-)(?:[0-9]){1,2}(|-)(?:[A-Za-z]){2}(|-)([0-9]){1,4})|(([A-Za-z]){2,3}(|-)([0-9]){1,4})/g;
    if (vRegex.test(vehicleNumber) && endDate > startDate) {
      mySqlConnection.query(
        "INSERT INTO dispatch(source_code,destination_code,transporter_code,vehicle_number,start_date,end_date) VALUES(?,?,?,?,?,?)",
        [
          sourceCode,
          destinationCode,
          transporterCode,
          vehicleNumber,
          startDate,
          endDate,
        ],
        (err, result) => {
          if (!err) {
            res.status(200).send({
              status: 200,
              message: "Created dispatch details successfully",
            });
          } else {
            res.status(400).send({
              message: "Bad Request",
            });
          }
        }
      );
    } else {
      res.status(400).send({
        message: "Please enter Valid details",
      });
    }
  } catch (e) {
    next(e);
  }
});

//for getting the disptahc details
router.get("/dispatch-details", (req, res, next) => {
  try {
    mySqlConnection.query(
      "SELECT * FROM dispatch order by start_date DESC",
      (err, result) => {
        if (!err) {
          if (!result) {
            res.status(200).send({
              message: "No Data Found, Please create some dispatch details",
            });
          } else {
            res.status(200).send({
              status: 200,
              data: result,
              message: "Retrieved all details successfully",
            });
          }
        }
      }
    );
  } catch (e) {
    next(e);
  }
});

//for getting the dispatch details in sorted manner
router.post("/sort-by-date", (req, res, next) => {
  try {
    if (req.body.order === "desc") {
      mySqlConnection.query(
        "SELECT * FROM dispatch order by start_date DESC",
        (err, result) => {
          if (!err) {
            if (!result) {
              res.status(200).send({
                message: "No Data Found, Please create some dispatch details",
              });
            } else {
              res.status(200).send({
                status: 200,
                data: result,
                message: "Retrieved all details successfully",
              });
            }
          }
        }
      );
    } else if (req.body.order === "asc") {
      mySqlConnection.query(
        "SELECT * FROM dispatch order by start_date ASC",
        (err, result) => {
          if (!err) {
            if (!result) {
              res.status(200).send({
                message: "No Data Found, Please create some dispatch details",
              });
            } else {
              res.status(200).send({
                status: 200,
                data: result,
                message: "Retrieved all details successfully",
              });
            }
          }
        }
      );
    }
  } catch (e) {
    next(e);
  }
});

//for searching based on source code
router.post("/search", (req, res, next) => {
  try {
    mySqlConnection.query(
      `SELECT * FROM dispatch WHERE source_code like '${req.body.search}%' or source_code like '%${req.body.search}' limit 2`,
      (err, result) => {
        if (!err) {
          if (result.length != 0) {
            res.status(200).send({
              status: 200,
              data: result,
              message: "Searched required result successfully",
            });
          } else {
            res.status(404).send({
              status: 404,
              message: "No result found",
            });
          }
        } else {
          res.status(400).send({
            status: 400,
            message: "Bad request",
          });
        }
      }
    );
  } catch (e) {
    next(e);
  }
});

//for creating excel sheet
router.get("/get-excel", (req, res, next) => {
  try {
    // Create an empty Excel object:
    let xlsx = officegen("xlsx");
    // Officegen calling this function after finishing to generate the xlsx document:
    xlsx.on("finalize", function (written) {
      console.log("Finish to create a Microsoft Excel document.");
    });
    // Officegen calling this function to report errors:
    xlsx.on("error", function (err) {
      console.log(err);
    });
    let sheet = xlsx.makeNewSheet();
    sheet.name = "Officegen Excel";
    sheet.data[0] = [];
    sheet.data[0][0] = "sr No.";
    sheet.data[0][1] = "Delivery No.";

    sheet.data[0][2] = "Source code";

    sheet.data[0][3] = "Destination Code";

    sheet.data[0][4] = "Transporter Code";

    sheet.data[0][5] = "Vehicle Number";

    sheet.data[0][6] = "Start Date";

    sheet.data[0][7] = "End date";
    let k = 1;
    mySqlConnection.query("SELECT * FROM dispatch", (err, result) => {
      if (!err) {
        for (var key in result) {
          sheet.data[k] = [];
          var current_row = result[key];

          // console.log(objSize);
          sheet.data[k][0] = k;
          if (current_row.hasOwnProperty("delivery_number")) {
            sheet.data[k][1] = current_row.delivery_number;
          }
          if (current_row.hasOwnProperty("source_code")) {
            sheet.data[k][2] = current_row.source_code;
          }
          if (current_row.hasOwnProperty("destination_code")) {
            sheet.data[k][3] = current_row.destination_code;
          }
          if (current_row.hasOwnProperty("transporter_code")) {
            sheet.data[k][4] = current_row.transporter_code;
          }
          if (current_row.hasOwnProperty("vehicle_number")) {
            sheet.data[k][5] = current_row.vehicle_number;
          }

          if (current_row.hasOwnProperty("start_date")) {
            sheet.data[k][6] = new Date(current_row.start_date)
              .toISOString()
              .slice(0, 10);
            // console.log("sdate",);
          }
          if (current_row.hasOwnProperty("end_date")) {
            // console.log("edate");
            sheet.data[k][7] = new Date(current_row.start_date)
              .toISOString()
              .slice(0, 10);
          }

          k = k + 1;
        }
        // Let's generate the Excel document into a file:

        let out = fs.createWriteStream("dispatch_report.xlsx");

        out.on("error", function (err) {
          console.log(err);
        });

        // Async call to generate the output file:
        xlsx.generate(out);

        var location = path.join(__dirname, "../../dispatch_report.xlsx");
        if (location) {
          res.status(200).send({
            status: 200,
            message:
              "The file has been downloaded to following directory - server/dispatch_report.xlsx",
          });
        }
      } else {
        res.status(400).send({
          message: "Cannot fetch details",
        });
      }
    });
  } catch (e) {
    next(e);
  }
});

module.exports = router;
