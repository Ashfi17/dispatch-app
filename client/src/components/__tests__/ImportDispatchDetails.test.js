import React from "react";
import { shallow } from "enzyme";
import { ImportDispatchDetails } from "../ImportDispatchDetails";
import { exportExcel } from "../../actions/Dispatch";

const addNewSDT = jest.fn();
const getSourceCode = jest.fn();
const getDestinationCode = jest.fn();
const getTransporterCode = jest.fn();
const addNewDispatch = jest.fn();
const history = {
  push: jest.fn(),
  // location: { state: { message: "Token Expired please login again" } },
};

const sourceCode = [{ source_code: "S001" }, { source_code: "S002" }];
const destinationCode = [{ destn_code: "D001" }, { destn_code: "D002" }];
const transporterCode = [
  { transporter_code: "T001" },
  { transporter_code: "T002" },
];

const wrapper = shallow(
  <ImportDispatchDetails
    addNewSDT={addNewSDT}
    getSourceCode={getSourceCode}
    getDestinationCode={getDestinationCode}
    getTransporterCode={getTransporterCode}
    addNewDispatch={addNewDispatch}
    history={history}
    sourceCode={sourceCode}
    destinationCode={destinationCode}
    transporterCode={transporterCode}
  />
);

describe("testing import dispatch details component", () => {
  it("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should test if the new source destination and transporter are getting added and the functions :OnTextChange,saveData are working ", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const saveData = jest.spyOn(wrapper.instance(), "saveData");
    let source = {
      target: {
        name: "source",
        value: "source name",
      },
    };
    let destination = {
      target: {
        name: "destination",
        value: "destination name",
      },
    };
    let transporter = {
      target: {
        name: "transporter",
        value: "transporter name",
      },
    };
    wrapper.find("#source-id").simulate("change", source);
    wrapper.find("#destination-id").simulate("change", destination);
    wrapper.find("#transporter-id").simulate("change", transporter);
    expect(wrapper.state().source).toBe("source name");
    expect(wrapper.state().destination).toBe("destination name");
    expect(wrapper.state().transporter).toBe("transporter name");
    expect(OnTextChange).toBeCalled();
    wrapper.find("#add-details-submit-button").simulate("click");
    expect(saveData).toBeCalled();
    expect(addNewSDT).toBeCalled();
  });
  it("should test if the functions handleFocus,addCode,handleBlur are working and if the source code are being added", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const handleFocus = jest.spyOn(wrapper.instance(), "handleFocus");
    const addCode = jest.spyOn(wrapper.instance(), "addCode");
    //for source code input field
    let sourceValue = {
      target: {
        name: "sourceValue",
        value: "sourceValue name",
      },
    };
    wrapper.find("#sourceValue-id").simulate("change", sourceValue);
    expect(OnTextChange).toBeCalled();
    wrapper.find("#sourceValue-id").simulate("focus", sourceValue);
    expect(handleFocus).toBeCalled();

    expect(getSourceCode).toBeCalled();
    expect(wrapper.state().sourceValueFlag).toBe(true);
    expect(wrapper.find("#source-item0").props().children).toBe("S001");
    wrapper.find("#source-item0").simulate("click");
    expect(addCode).toBeCalled();
    expect(wrapper.state().sourceValueFlag).toBe(false);
  });
  it("should test if the functions handleFocus,addCode,handleBlur are working and if the destination code are being added", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const handleFocus = jest.spyOn(wrapper.instance(), "handleFocus");
    const addCode = jest.spyOn(wrapper.instance(), "addCode");

    //for destination code input field
    let destinationValue = {
      target: {
        name: "destinationValue",
        value: "destinationValue name",
      },
    };
    wrapper.find("#destinationValue-id").simulate("change", destinationValue);
    expect(OnTextChange).toBeCalled();
    wrapper.find("#destinationValue-id").simulate("focus", destinationValue);
    expect(handleFocus).toBeCalled();
    expect(getDestinationCode).toBeCalled();
    expect(wrapper.state().destinationValueFlag).toBe(true);
    expect(wrapper.find("#destination-item0").props().children).toBe("D001");
    wrapper.find("#destination-item0").simulate("click");
    expect(addCode).toBeCalled();
    expect(wrapper.state().destinationValueFlag).toBe(false);
  });
  it("should test if the functions handleFocus,addCode,handleBlur are working and if the transporter code are being added", () => {
    const OnTextChange = jest.spyOn(wrapper.instance(), "OnTextChange");
    const handleFocus = jest.spyOn(wrapper.instance(), "handleFocus");
    const addCode = jest.spyOn(wrapper.instance(), "addCode");

    //for transporter code input field
    let transporterValue = {
      target: {
        name: "transporterValue",
        value: "transporterValue name",
      },
    };
    wrapper.find("#transporterValue-id").simulate("change", transporterValue);
    expect(OnTextChange).toBeCalled();
    wrapper.find("#transporterValue-id").simulate("focus", transporterValue);
    expect(handleFocus).toBeCalled();
    expect(getTransporterCode).toBeCalled();
    expect(wrapper.state().transporterValueFlag).toBe(true);
    expect(wrapper.find("#transporter-item0").props().children).toBe("T001");
    wrapper.find("#transporter-item0").simulate("click");
    expect(addCode).toBeCalled();
    expect(wrapper.state().transporterValueFlag).toBe(false);
  });
});
