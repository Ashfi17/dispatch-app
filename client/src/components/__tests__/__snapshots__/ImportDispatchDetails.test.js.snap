// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing import dispatch details component should render the component 1`] = `
<div>
  <div
    className="add-source-destination"
  >
    <h1>
      Add new source,destination and transporter
    </h1>
    <input
      className="data-input-field"
      id="source-id"
      name="source"
      onChange={[Function]}
      placeholder="source name"
      type="text"
      value=""
    />
    <input
      className="data-input-field"
      id="destination-id"
      name="destination"
      onChange={[Function]}
      placeholder="destination name"
      type="text"
      value=""
    />
    <input
      className="data-input-field"
      id="transporter-id"
      name="transporter"
      onChange={[Function]}
      placeholder="transporter name"
      type="text"
      value=""
    />
    <button
      className="submit-button"
      id="add-details-submit-button"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
  <h1
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    Add New Dispatch Details
  </h1>
  <div
    className="add-new-dispatch-info"
  >
    <div
      className="left-card-section"
    >
      <input
        className="data-input-field"
        id="sourceValue-id"
        name="sourceValue"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Source code eg:S001"
        type="text"
        value=""
      />
      <input
        className="data-input-field right-section-input"
        id="destinationValue-id"
        name="destinationValue"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Destination code eg:D001"
        type="text"
        value=""
      />
      <input
        className="data-input-field right-section-input"
        id="transporterValue-id"
        name="transporterValue"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Transporter code eg:T001"
        type="text"
        value=""
      />
    </div>
    <div
      className="right-card-section"
    >
      <input
        className="data-input-field "
        id="vehicleNumber-id"
        name="vehicleNumber"
        onChange={[Function]}
        placeholder="vehicle number"
        type="text"
        value=""
      />
      <input
        className="data-input-field right-section-input"
        id="startDate-id"
        name="startDate"
        onChange={[Function]}
        placeholder="start Date"
        type="text"
        value=""
      />
      <input
        className="data-input-field right-section-input"
        id="endDate-id"
        name="endDate"
        onChange={[Function]}
        placeholder="end Date"
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    className="save-dispatch-button"
    id="add-dispatch-details-button"
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
